README.md
# EEL - A simple Proxy Service for JSON Event Transformation and Forwarding

Requires go 1.7 or greater.

----
It's simple - a single JSON event comes in and one (or maybe a few) transformed events get out.
Events are arbitrary JSON encoded payloads and they are typically sent around as
HTTP POSTs. EEL is stateless and therefore scales easily.

![](doc/eel.png)

In this example we removed the unwanted "detail" field from the event and adjusted
the schema slightly.

Converting JSON payload A from service S1 into JSON payload B for service S2 is a very common
operation we perform all the time in service oriented architectures. Here are some good reasons
for event transformation:

1. REDUCTION: Remove everything that's not needed and only pass the data actually required by the downstream service. This can provide performance benefits for the downstream service if processing is expensive and events are bloated.
1. CANONICALIZATION / CLEANUP: Adjust the structure of the JSON event to validate against a given schema. Typically this can be achieved by rearranging elements in the JSON event, renaming keys and possibly filtering out unwanted subsets of data.
1. META-TAGGING: Inject additional metadata into the JSON event. Examples are injecting UUIDs or timestamps (for example for log tracing), tagging or labeling events with `type` fields, calling external auxiliary services and injecting their output into the downstream payload (provided it is JSON).
1. MAPPING / SUBSTITUTION: Map one ID type to another via look up using an external lookup service.
1. FILTERING: Filter out unwanted events to reduce load for downstream services.
1. FANOUT: Forward a single incoming JSON event to multiple downstream services, either by applying the same transformation for all services or by performing different transformations.

You could write each of these transformations in just a few lines of go code, and we often do.
The downside is that transformations become engraved in code and cannot be changed easily.

EEL is offering a simple JSON template language combined with a proxy micro service to relay
and transform events between upstream and downstream services. The goals of the EEL template language are to be
simple and yet powerful. EEL can be the glue in a JSON based service oriented eco-system.

The syntax of the EEL transformation language is inspired by a few of the core concepts from XPath and XSLT applied to JSON.

## Installation

```
go get -u github.com/comcast/eel
```

## Usage

```
./eel [options...]
```

Options:

```
-config  path to config.json (default is ./config-eel/config.json)
-handlers  path to handlers (default is ./config-handlers)
-loglevel  log level (default is "info")
-env  environment name such as qa, prod for logging (default is "default")
```

## No Nonsense

```
go build -o bin/eel
./bin/starteel.sh
```

## Docker Alternative

Docker 1.13+ verified.

Build a dev image

```
docker build -t eel:dev .
```

Run an instance

```
docker run --rm -p 8080:8080 --name eel-dev eel:dev
```

The command above will utilize port `8080` of your host.
You can change it to any other port via `-p ANYOTHERPORT:8080`

To pass parameters to `eel` you can use `EEL_PARAMS` env variable, e.g.
```
docker run --rm -e "EEL_PARAMS=-loglevel error" -p 8080:8080 --name eel-dev eel:dev
```

## A Simple Example

Transformation handlers are used to tell EEL if and how to process JSON events it receives and where to
forward them to. Each transformation handler is stored in a single JSON file and there may be several such
handler files each of them taking care of a certain class of events. See [here](doc/handlers.md) for more details.

Edit the default transformation handler in [config-handlers/tenant1/default.json](config-handlers/tenant1/default.json)
so it looks like this (there's only a small change needed in the `Transformation` section).

```
{
    "Version": "1.0",
    "Name": "Default",
    "Info": "",
    "Active": true,
    "Match": null,
    "IsMatchByExample": false,
    "TerminateOnMatch": true,
    "Transformation": {
        "{{/event}}": "{{/}}"
    },
    "IsTransformationByExample": false,
    "Path": "",
    "Verb": "POST",
    "Endpoint": "http://localhost:8082",
    "HttpHeaders": {
      "X-B3-TraceId": "{{traceid()}}",
      "X-Tenant-Id": "{{tenant()}}"
    }
}
```

The center piece of the handler configuration is the `Transformation` section which uses [JPath](doc/jpath.md) expressions to describe the structural transformations to be performed on the incoming event before forwarding it to the downstream service. In this example we are telling EEL to take the entire payload unmodified and wrap it inside a new element called `event`.

Compile EEL.

EEL is implemented in go and you will need a [golang](https://golang.org/) environment to compile the code. However, you don't need any go coding skills to understand and author EEL templates!

The make file and shell scripts have been tested on Mac and Linux environments.

```
make all
```

Launch EEL as a service and start listening to incoming events.

```
./bin/starteel.sh
```

Send a test event to EEL.

```
curl -X POST --data '{ "message" : "hello world!!!" }' http://localhost:8080/v1/events
```

Output:

```
{"status":"processed"}
```

You can check EEL's log output `eel.log` to see if and how the event is processed. You can also ask for a detailed
debug response using the `X-Debug` header.

```
curl -X POST -H 'X-Debug: true' --data '{ "message" : "hello world!!!" }' http://localhost:8080/v1/events
```

Output:

```
[
	{
		"api": "http",
		"handler": "Default",
		"tenant.id": "tenant1",
		"trace.in.data": {
			"message": "hello world!!!"
		},
		"trace.out.data": {
			"event": {
				"message": "hello world!!!"
			}
		},
		"trace.out.endpoint": "http://localhost:8088",
		"trace.out.headers": {
			"X-B3-TraceId": "20073ee4-d681-4ab5-a973-50c978cd1111",
			"X-Tenant-Id": "tenant1"
		},
		"trace.out.path": "",
		"trace.out.protocol": "http",
		"trace.out.url": "http://localhost:8088",
		"trace.out.verb": "POST",
		"tx.id": "20073ee4-d681-4ab5-a973-50c978cd1111",
		"tx.traceId": "20073ee4-d681-4ab5-a973-50c978cd1111"
	}
]
```

Or, you can use the `/v1/sync/events` instead of the `/v1/events` endpoint to get an immediate response containing the transformed event as it would be forwarded to the downstream service.

```
curl -X POST --data '{ "message" : "hello world!!!" }' http://localhost:8080/v1/sync/events
```

Output:

```
{
	"event": {
		"message": "hello world!!!"
	}
}
```

To review the current set of active transformation handlers call the health check API.

```
curl http://localhost:8080/v1/health
```

Stop EEL.

```
./bin/stopeel.sh
```

## EEL as Command Line Tool

You can also start experimenting with EEL by using the command line parameters. Example:

```
./eel -in='{"foo":"bar"}' -tf='{"Foo":"{{/foo}}"}' -istbe=true
```

More examples can be found [here ](doc/cmd.md).


## Exploring EEL Features

The unit tests are a good starting point to learn more about EEL features and look at some examples in detail.

Each test is contained in its own configuration folder, for example `eel/test/data/test01`. Each test folder
contains a complete set of handler configurations for EEL in the `handlers` subfolder (usually just one handler),
an example input event `in.json` and one or more expected output events `out.json`.

The tests can be executed using `go test`:

```
cd eel/test
go test -v
```

Or, you can launch EEL with the handler configurations for a specific test and send the `in.json` event manually.

```
./eel -handlers=test/data/test01/handlers > eel.log &
curl -X POST --data @test/data/test01/in.json http://localhost:8080/v1/sync/events
```

No | Name | Test Name | Description
--- |--- | --- | ---
0 | Identity Transformation | [TestDontTouchEvent](test/data/test00) | Doesn't apply any transformation and forwards everything unchanged.
1 | Canonicalize | [TestCanonicalizeEvent](test/data/test01) | Simple structural changes and array path selectors.
2 | External Lookup | [TestInjectExternalServiceResponse](test/data/test02) | Get JSON data from external service and inject into payload.
3 | Transformation By Example | [TestTransformationByExample](test/data/test03) | Describe transformation using by-example syntax rather than using by-path syntax.
4 | Named Transformations | [TestNamedTransformations](test/data/test04) | Choose from several named transformations based on input data.
5 | Conditional Message Generation | [TestMessageGeneration](test/data/test05) | Assemble message string with ifte() and equals().
6 | Handler Matching 1 | [TestTerminateOnMatchTrue](test/data/test06) | Pick best matching handler and forward single event.
7 | Handler Matching 2 | [TestTerminateOnMatchFalse](test/data/test07) | Pick all matching handlers and forward multiple events.
8 | Multi Tenancy | [TestMultiTenancy](test/data/test08) | Handlers for different tenants or apps.
9 | Cascade | [TestSequentialHandlerCascade](test/data/test09) | Cascade of multiple handlers which will be executed sequentially by using EEL recursively.
10 | Java Script For Everything Else | [TestJavaScript](test/data/test10) | If you really can't avoid it, resort to Java Script.
11 | Handler Matching 3 | [TestMatchByExample](test/data/test11) | Matching handlers using by-example syntax.
12 | Convert Headers To Payload | [TestHeaders](test/data/test12) | Inject HTTP headers from upstream service into JSON event for downstream service.
13 | Custom Properties | [TestCustomProperties](test/data/test13) | Custom properties in handlers for sharing data.
14 | Fan Out | [TestFanOut](test/data/test14) | Send incoming event to several downstream services.
15 | Basic String Operations | [TestStringOps](test/data/test15) | Uppercase, lowercase, substring.
16 | Named Transformations 2 | [TestNamedTransformations2](test/data/test16) | Perform named transformation on external document.
17 | Contains | [TestContains](test/data/test17) | Check if one JSON event is contained in another.
18 | Tenant Id Header | [TestMultiTenency2](test/data/test18) | Pass in tenant id as HTTP header.
19 | Conditional Message Generation 2 | [TestMessageGeneration2](test/data/test19) | Use case() function to simplify conditional string generation.
20 | Regex | [TestRegex](test/data/test20) | Use regex() to evaluate regular expressions.
22 | Filter By Path | [TestFilterByPath](test/data/test22) | Filter event after transformation using by-path syntax.
23 | Filter By Example | [TestFilterByExample](test/data/test23) | Filter event after transformation using by-example syntax.
25 | Array Path Selector | [TestArrayPathSelector](test/data/test25) | Select elements from arrays by index or by path.
27 | Iterate Over Array | [NamedTransformationsAndArrays](test/data/test27) | Iterate over array and apply named transformation.
32 | Simple Types | [NamedTransformationsAndSimpleTypes](test/data/test32) | Apply named transformation to a simple type.
34 | Join | [Join](test/data/test34) | Merge two JSON documents.
41 | Partial Matching | [MatchPartialArrays3](test/data/test41) | Match event against partial pattern.
46 | Filter Cascade | [FilterCascade](test/data/test46) | Apply multiple filters.
47 | Choose From Array | [ChooseFromArray](test/data/test47) | Choose elements from array by pattern.
48 | Named Transformation With Array And Pattern | [NamedTransformationWithArrayAndPattern](test/data/test48) | Apply named transformation by pattern.
49 | Named Transformation With Array And Join | [NamedTransformationWithArrayAndJoin](test/data/test49) | Apply named transformation with join.
51 | Complex Example | [ComplexExample](test/data/test51) | A real world example.
53 | Crush | [Crush](test/data/test53) | Flatten a deeply nested array.

## Further Reading

* [JSON Path Expressions](doc/jpath.md)
* [Transformation Handler Configuration](doc/handlers.md)
* [Function Reference](doc/functions.md)
* [Global Configuration Parameters](doc/configuration.md)
* [Debug Tools](doc/debug.md)
* [EEL Web API Reference](doc/apis.md)
* [EEL Transformations as Go Library](doc/lib.md)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
This `CHANGELOG.md` implements the spirit of http://keepachangelog.com/.

## [1.43](https://github.com/Comcast/eel/compare/v1.42.0...dev) - [Unreleased]

### Fixed
* XRULES-19652: panic in nae

## [1.42](https://github.com/Comcast/eel/compare/v1.41.0...v1.42.0) - 2022-03-28

### Added
* Add metrics and tracing to eel

## [1.41](https://github.com/Comcast/eel/compare/v1.40.0...v1.41.0) - 2021-06-02

### Added
* Logging for partner Id
* Covert to go mod

## [1.40](https://github.com/Comcast/eel/compare/v1.39.0...v1.40.0) - 2021-05-10

### Added
* AllowPartner config parameter
* partner() function
* Add logic to handle http partner header

## [1.39](https://github.com/Comcast/eel/compare/v1.38.0...v1.39.0) - 2020-12-04

## [1.38](https://github.com/Comcast/eel/compare/v1.37.0...v1.38.0) - 2020-11-16

### Added
* XRULES-20200606: Add function 'param' to read query string parameters of incoming event url
* XRULES-20200606: Add use of HTTP_PROXY environment variable to send data via a proxy

### Updated
* XRULES-20200606: Docker build; Make smaller runtime image by using multi-stage build, and enable testing

## [1.37](https://github.com/Comcast/eel/compare/v1.36.0...v1.37.0) - 2020-03-09

### Added
* XRULES-16187: Pt 2 - Implement: Isolate flow executions in masheens

## [1.36](https://github.com/Comcast/eel/compare/v1.35.0...v1.36.0) - 2019-11-25

## [1.35](https://github.com/Comcast/eel/compare/v1.34.0...v1.35.0) - 2019-11-04

### Added
* XRULES-15127: Updates to Profile Service Endpoint

## [1.34](https://github.com/Comcast/eel/compare/v1.33.0...v1.34.0) - 2019-10-17

### Added
* Ability to specify a basepath in the URL

## [1.33](https://github.com/Comcast/eel/compare/v1.32.0...v1.33.0) - 2019-09-23

## [1.32](https://github.com/Comcast/eel/compare/v1.31.0...v1.32.0) - 2019-08-09

## [1.31](https://github.com/Comcast/eel/compare/v1.30.0...v1.31.0) - 2019-07-19

### Added
* Ability to send abitrary parameters to a pubblisher
* Remove kafka/asyncReply related parameters

## [1.30](https://github.com/Comcast/eel/compare/v1.29...v1.30.0) - 2019-06-07
* XRULES-13840: Default handlers in EEL
* XRULES-13983: Flow execution log issues

## [1.29](https://github.com/Comcast/eel/compare/v1.28...v1.29) - 2019-05-17

## [1.28](https://github.com/Comcast/eel/compare/v1.27...v1.28) - 2019-04-29

## [1.27](https://github.com/Comcast/eel/compare/v1.26...v1.27) - 2019-04-08

## [1.26](https://github.com/Comcast/eel/compare/v1.25...v1.26) - 2019-03-18

## [1.25](https://github.com/Comcast/eel/compare/v1.24...v1.25) - 2019-02-15

### Added
* Two new EEL functions: propExists and toTS

## [1.24](https://github.com/Comcast/eel/compare/v1.23...dev) - 2019-01-28

## Updated
* XRULES-12619: All components need to log Gears Portal application id

## [1.23](https://github.com/Comcast/eel/compare/v1.22...v1.23) - 2019-01-07

### Updated
* reduce log lines

### Added
* Add a global flag to disable all plugins except webhook

## [1.22](https://github.com/Comcast/eel/compare/v1.21...v1.22) - 2018-11-12

### Added
* Add function base64decode
* Add function loadFile
* Add function hmac
* Add optional multiple partitions for kafka publisher

## [1.21](https://github.com/Comcast/eel/compare/v1.20...v1.21) - 2018-07-09

## [1.20](https://github.com/Comcast/eel/compare/v1.19...v1.20) - 2018-06-18

## [1.19](https://github.com/Comcast/eel/compare/v1.18...v1.19) - 2018-05-24

## [1.18](https://github.com/Comcast/eel/compare/v1.17...v1.18) - 2018-05-07

## [1.17](https://github.com/Comcast/eel/compare/v1.16...v1.17) - 2018-03-26

### Added
* Gears suppport

## [1.16](https://github.com/Comcast/eel/compare/v1.15...v1.16) - 2018-03-05

### Added
* Added a new config parameter: ElementsAuth

### Fixed
* XRULES-10493: support escape character in function parameters

## [1.15](https://github.com/Comcast/eel/compare/v1.14...v1.15) - 2017-12-01

### Fixed
* Update retry to not retry for 300 and 400

## [1.14](https://github.com/Comcast/eel/compare/v1.13...v1.14) - 2017-11-08

### Fixed
* Fixed bad urls in `CHANGELOG.md`

### Changed
* Commented out array_length_error log to address XRULES-9749

## [1.13](https://github.com/Comcast/eel/compare/v1.12...v1.13) - [2017-09-08]

### Changed
* add tenantId in logs

## [1.12](https://github.com/Comcast/eel/compare/v1.11...v1.12) - [2017-07-28]

### Fixed
* XRULES-8923: erroneous metric on published events

## [1.11](https://github.com/Comcast/eel/compare/v1.10...v1.11) - [2017-06-19]

### Added
* This `CHANGELOG.md` file
* .github/PULL_REQUEST_TEMPLATE.md to improve our PR process

### Fixed
* XRULES-8388: Panic handling code caused the code to panic


main.go
/**
 * Copyright 2015 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// A simple proxy service to forward JSON events and transform or filter them along the way.
package main

import (
	"flag"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"time"

	. "github.com/Comcast/eel/jtl"
	. "github.com/Comcast/eel/util"

	_ "net/http/pprof"
)

// build hint: go build -ldflags "-X main.Version 2.0"

var (
	Version = "1.0"
)

var (
	// proxy params
	env         = flag.String("env", "default", "environment name such as qa, prod for logging")
	basePath    = flag.String("path", "", "base path for config.json and handlers (optional)")
	configPath  = flag.String("config", "", "path to config.json (optional)")
	handlerPath = flag.String("handlers", "", "path to handlers (optional)")
	logLevel    = flag.String("loglevel", L_InfoLevel, "log level (optional)")
	// cmd params
	in    = flag.String("in", "", "incoming event string or @file")
	tf    = flag.String("tf", "", "transformation string or @file")
	istbe = flag.Bool("istbe", true, "is template by example flag")
)

// useCores if GOMAXPROCS not set use all cores you got.
func useCores(ctx Context) {
	cores := os.Getenv("GOMAXPROCS")
	if cores == "" {
		n := runtime.NumCPU()
		ctx.Log().Info("action", "use_cores", "cores", n)
		runtime.GOMAXPROCS(n)
		cores = strconv.Itoa(n)
	} else {
		ctx.Log().Info("action", "use_cores_from_env", "cores", cores)
	}
}

// initLogging sets up context and stats loop.
func initLogging() {
	if *basePath != "" {
		BasePath = *basePath
	}
	if *configPath != "" {
		ConfigPath = filepath.Join(BasePath, *configPath)
	} else {
		ConfigPath = filepath.Join(BasePath, EelConfigFile)
	}
	LogLevel = *logLevel
	InitGctx(*logLevel)
	config := GetConfigFromFile(Gctx)
	if *handlerPath != "" {
		HandlerPath = *handlerPath
	} else if config.HandlerConfigPath != "" {
		HandlerPath = config.HandlerConfigPath
	}
	AppId = config.AppName
	Gctx.AddLogValue("app.id", AppId)
	InstanceName, _ = os.Hostname()
	Gctx.AddLogValue("instance.id", InstanceName)
	if *env != "" {
		EnvName = *env
		Gctx.AddLogValue("env.name", EnvName)
	}
	Gctx.AddValue(EelStartTime, time.Now().Local().Format("2006-01-02 15:04:05 +0800"))
	stats := new(ServiceStats)
	Gctx.AddValue(EelTotalStats, stats)
	Gctx.AddValue(Eel1MinStats, new(ServiceStats))
	Gctx.AddValue(Eel5MinStats, new(ServiceStats))
	Gctx.AddValue(Eel1hrStats, new(ServiceStats))
	Gctx.AddValue(Eel24hrStats, new(ServiceStats))

	Gctx.AddConfigValue(EelTraceLogger, NewTraceLogger(Gctx, config))

	getWorkQueueFillLevel := func(tenantId string) int {
		wd := GetWorkDispatcher(Gctx, tenantId)
		if wd != nil {
			return len(wd.WorkQueue)
		}
		return -1
	}

	getNumWorkersIdle := func(tenantId string) int {
		wd := GetWorkDispatcher(Gctx, tenantId)
		if wd != nil {
			return len(wd.WorkerQueue)
		}
		return -1
	}

	tenantIds := make([]string, 0, len(config.WorkerPoolSize))
	for k := range config.WorkerPoolSize {
		tenantIds = append(tenantIds, k)
	}
	//Gctx.Log().Debug("tenantIds", tenantIds)
	Gctx.AddValue(EelTenantIds, tenantIds)

	if config.LogStats {
		go Gctx.Log().RuntimeLogLoop(time.Duration(60)*time.Second, -1)
		for _, tenantId := range tenantIds {
			go stats.StatsLoop(Gctx, 300*time.Second, -1, Eel5MinStats, getWorkQueueFillLevel, getNumWorkersIdle, tenantId)
			go stats.StatsLoop(Gctx, 60*time.Second, -1, Eel1MinStats, getWorkQueueFillLevel, getNumWorkersIdle, tenantId)
			go stats.StatsLoop(Gctx, 60*time.Minute, -1, Eel1hrStats, getWorkQueueFillLevel, getNumWorkersIdle, tenantId)
			go stats.StatsLoop(Gctx, 24*time.Hour, -1, Eel24hrStats, getWorkQueueFillLevel, getNumWorkersIdle, tenantId)
		}
	}
}

func registerAdminServices() {
	c := Gctx.SubContext()
	// old handlers
	http.HandleFunc("/health/shallow", c.WrapPanicHttpHandler(NilHandler))
	http.HandleFunc("/health/deep", c.WrapPanicHttpHandler(StatusHandler))
	http.HandleFunc("/health", c.WrapPanicHttpHandler(StatusHandler))
	http.HandleFunc("/status", c.WrapPanicHttpHandler(StatusHandler))
	http.HandleFunc("/pluginconfigs", c.WrapPanicHttpHandler(PluginConfigHandler))
	http.HandleFunc("/plugins", c.WrapPanicHttpHandler(ManagePluginsUIHandler))
	http.HandleFunc("/plugins/", c.WrapPanicHttpHandler(ManagePluginsHandler))
	http.HandleFunc("/reload", c.WrapPanicHttpHandler(ReloadConfigHandler))
	http.HandleFunc("/toggletracelogger", c.WrapPanicHttpHandler(TraceLogConfigHandler))
	http.HandleFunc("/vet", c.WrapPanicHttpHandler(VetHandler))
	http.HandleFunc("/version", c.WrapPanicHttpHandler(VersionHandler))
	http.HandleFunc("/test", c.WrapPanicHttpHandler(TopicTestHandler))
	http.HandleFunc("/test/handlers", c.WrapPanicHttpHandler(HandlersTestHandler))
	http.HandleFunc("/test/process/", c.WrapPanicHttpHandler(ProcessExpressionHandler))
	http.HandleFunc("/test/ast", c.WrapPanicHttpHandler(ParserDebugHandler))
	http.HandleFunc("/test/astjson/", c.WrapPanicHttpHandler(GetASTJsonHandler))
	http.HandleFunc("/test/asttree/", c.WrapPanicHttpHandler(ParserDebugVizHandler))
	http.HandleFunc("/event/dummy", c.WrapPanicHttpHandler(DummyEventHandler))
	// v1 handlers
	http.HandleFunc("/v1/health/shallow", c.WrapPanicHttpHandler(NilHandler))
	http.HandleFunc("/v1/health/deep", c.WrapPanicHttpHandler(StatusHandler))
	http.HandleFunc("/v1/health", c.WrapPanicHttpHandler(StatusHandler))
	http.HandleFunc("/v1/status", c.WrapPanicHttpHandler(StatusHandler))
	http.HandleFunc("/v1/pluginconfigs", c.WrapPanicHttpHandler(PluginConfigHandler))
	http.HandleFunc("/v1/plugins", c.WrapPanicHttpHandler(ManagePluginsUIHandler))
	http.HandleFunc("/v1/plugins/", c.WrapPanicHttpHandler(ManagePluginsHandler))
	http.HandleFunc("/v1/reload", c.WrapPanicHttpHandler(ReloadConfigHandler))
	http.HandleFunc("/v1/toggletracelogger", c.WrapPanicHttpHandler(TraceLogConfigHandler))
	http.HandleFunc("/v1/vet", c.WrapPanicHttpHandler(VetHandler))
	http.HandleFunc("/v1/version", c.WrapPanicHttpHandler(VersionHandler))
	http.HandleFunc("/v1/test", c.WrapPanicHttpHandler(TopicTestHandler))
	http.HandleFunc("/v1/test/handlers", c.WrapPanicHttpHandler(HandlersTestHandler))
	http.HandleFunc("/v1/test/process/", c.WrapPanicHttpHandler(ProcessExpressionHandler))
	http.HandleFunc("/v1/test/ast", c.WrapPanicHttpHandler(ParserDebugHandler))
	http.HandleFunc("/v1/test/astjson/", c.WrapPanicHttpHandler(GetASTJsonHandler))
	http.HandleFunc("/v1/test/asttree/", c.WrapPanicHttpHandler(ParserDebugVizHandler))
	http.HandleFunc("/v1/event/dummy", c.WrapPanicHttpHandler(DummyEventHandler))
	http.HandleFunc("/v1/event/panic", c.WrapPanicHttpHandler(PanicEventHandler))
	//
	http.Handle("/img/", http.StripPrefix("/img/", http.FileServer(http.Dir(filepath.Join(BasePath, "mascot")))))
}

func main() {
	flag.Parse()
	if *tf != "" {
		eelCmd(*in, *tf, *istbe)
	} else {
		initLogging()
		ReloadConfig()
		GetConfig(Gctx).Version = Version
		InitHttpTransport(Gctx)
		ctx := Gctx.SubContext()
		ctx.Log().Info("action", "starting", "version", Version)
		useCores(ctx)
		dc := NewLocalInMemoryDupChecker(GetConfig(ctx).DuplicateTimeout, 10000)
		Gctx.AddValue(EelDuplicateChecker, dc)

		tenantIds := Gctx.Value(EelTenantIds).([]string)
		for _, tenantId := range tenantIds {
			dp := NewWorkDispatcher(GetConfig(ctx).WorkerPoolSize[tenantId], GetConfig(ctx).MessageQueueDepth, tenantId)
			dp.Start(ctx)
			Gctx.AddValue(EelDispatcher+"_"+tenantId, dp)
			ctx.Log().Info("action", "start_worker_pool", "size", GetConfig(ctx).WorkerPoolSize[tenantId], "queue_depth", GetConfig(ctx).MessageQueueDepth, "tenant_id", tenantId)
		}
		registerAdminServices()
		// register inbound plugins
		RegisterInboundPluginType(NewStdinPlugin, "STDIN")
		RegisterInboundPluginType(NewWebhookPlugin, "WEBHOOK")
		LoadInboundPlugins(Gctx, true)
		// hang on channel forever
		<-make(chan int)
	}
}

//used to test  HttpHandlerFunc panic can be removed later
func PanicEventHandler(w http.ResponseWriter, r *http.Request) {
	ctx := Gctx.SubContext()
	ctx.Log().Info("op", "PanicEventNoHandler")
	panic("I am panic now and Nobody care!")
}
